packages/services/src/sortino-ratio.service.ts

```diff
@@ -0,0 +1,164 @@
+    this.logger = logger;
+  }
+
+  /**
+   * Calculate Sortino Ratio metrics (without persisting)
+   * Uses database-level window functions
+   *
+   * @param agentId Agent ID
+   * @param competitionId Competition ID
+   * @returns Calculated metrics
+   */
+  async calculateSortinoRatio(
+    agentId: string,
+    competitionId: string,
+  ): Promise<SortinoMetrics> {
+    try {
+      this.logger.info(
+        `[SortinoRatio] Calculating Sortino Ratio for agent ${agentId} in competition ${competitionId}`,
+      );
+
+      // 1. Get competition to validate it exists and has started
+      const competition = await this.competitionRepo.findById(competitionId);
+      if (!competition) {
+        throw new Error(`Competition ${competitionId} not found`);
+      }
+
+      if (!competition.startDate) {
+        throw new Error(`Competition ${competitionId} has not started yet`);
+      }
+
+      const endDate = competition.endDate
+        ? new Date(competition.endDate)
+        : new Date();
+
+      this.logger.debug(
+        `[SortinoRatio] Competition period: ${competition.startDate.toISOString()} to ${endDate.toISOString()}`,
+      );
+
+      // 2. Calculate all metrics at database level using window functions
+      const metrics = await this.competitionRepo.calculateSortinoMetricsSQL(
+        agentId,
+        competitionId,
+        this.MAR,
+      );
+
+      if (metrics.snapshotCount < 2) {
+        this.logger.warn(
+          `[SortinoRatio] Insufficient snapshots for agent ${agentId}: ${metrics.snapshotCount} found`,
+        );
+        throw new Error("Insufficient data: Need at least 2 snapshots");
+      }
+
+      // 3. Calculate Sortino Ratio from database metrics
+      const avgReturn = new Decimal(metrics.avgReturn);
+      const downsideDeviation = new Decimal(metrics.downsideDeviation);
+      const simpleReturn = new Decimal(metrics.simpleReturn);
+
+      const sortinoRatio = this.computeSortinoRatio(
+        avgReturn,
+        downsideDeviation,
+      );
+
+      this.logger.info(
+        `[SortinoRatio] Calculated metrics for agent ${agentId}: ` +
+          `Sortino=${sortinoRatio.toFixed(4)}, ` +
+          `Avg Return=${(avgReturn.toNumber() * 100).toFixed(2)}%, ` +
+          `Downside Dev=${(downsideDeviation.toNumber() * 100).toFixed(2)}%, ` +
+          `Simple Return=${(simpleReturn.toNumber() * 100).toFixed(2)}%`,
+      );
+
+      // Return calculated metrics
+      const calculatedMetrics: SortinoMetrics = {
+        agentId,
+        competitionId,
+        sortinoRatio: sortinoRatio.toFixed(8),
+        downsideDeviation: downsideDeviation.toFixed(8),
+        annualizedReturn: avgReturn.toFixed(8),
+        simpleReturn: simpleReturn.toFixed(8),
+        snapshotCount: metrics.snapshotCount,
+      };
+
+      return calculatedMetrics;
+    } catch (error) {
+      this.logger.error(
+        `[SortinoRatio] Error calculating Sortino Ratio for agent ${agentId}:`,
```

Comment:
`error({ error}, ...)`

Reply:
Fixed this one and what should be the remainder of the others here: [fa6bf73](https://github.com/recallnet/js-recall/pull/1498/commits/fa6bf730ddd1227b4eaf2a54904d9c4c53f3134a)